pkg/oidc/check/flow.go

```diff
@@ -0,0 +1,167 @@
+package check
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/pkg/browser"
+	"github.com/weaveworks/weave-gitops/pkg/logger"
+	"github.com/weaveworks/weave-gitops/pkg/server/auth"
+	"golang.org/x/oauth2"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Options struct {
+	ClientID                   string
+	ClientSecret               string
+	IssuerURL                  string
+	SecretName                 string
+	SecretNamespace            string
+	Scopes                     []string
+	ClaimUsername              string
+	OpenURL                    func(string) error
+	InsecureSkipSignatureCheck bool
+}
+
+type Claims struct {
+	Username string
+	Groups   []string
+}
+
+// GetClaims retrieves OIDC claims by sending the user through an authorization code flow. It spins
+// up a temporary web server, sets the server's address as redirect URI in the authentication request
+// and subsequently exchanges the authorization code for an ID token.
+func GetClaims(ctx context.Context, opts Options, log logger.Logger, c client.Client) (Claims, error) {
+	res := Claims{}
+	if opts.SecretName != "" {
+		if err := optsFromSecret(ctx, &opts, log, c); err != nil {
+			return res, fmt.Errorf("failed reading options from Secret: %w", err)
+		}
+	}
+
+	if opts.ClaimUsername == "" {
+		opts.ClaimUsername = auth.ClaimUsername
+	}
+
+	if len(opts.Scopes) == 0 {
+		opts.Scopes = auth.DefaultScopes
+	}
+
+	provider, err := oidc.NewProvider(ctx, opts.IssuerURL)
+	if err != nil {
+		return res, fmt.Errorf("could not create provider: %w", err)
+	}
+
+	oauth2Config := oauth2.Config{
+		ClientID:     opts.ClientID,
+		ClientSecret: opts.ClientSecret,
+		RedirectURL:  "http://localhost:9876",
+		Endpoint:     provider.Endpoint(),
+		Scopes:       opts.Scopes,
+	}
+
+	log.Waitingf("Opening browser. If this does not work, please open the following URL in your browser:\n")
+	authCodeURL := oauth2Config.AuthCodeURL("state", oauth2.AccessTypeOffline)
+	log.Println("%s\n", authCodeURL)
+	var openErr error
+	if opts.OpenURL != nil {
+		openErr = opts.OpenURL(authCodeURL)
+	} else {
+		openErr = browser.OpenURL(authCodeURL)
+	}
+	if openErr != nil {
+		log.Failuref("Failed to open browser: %s. You can still open the URL manually.", openErr)
+	}
+
+	verifier := provider.Verifier(&oidc.Config{
+		ClientID:                   opts.ClientID,
+		InsecureSkipSignatureCheck: opts.InsecureSkipSignatureCheck,
+	})
+	claims, err := retrieveClaims(log, oauth2Config, verifier)
+	if err != nil {
+		return res, fmt.Errorf("failed retrieving claims: %w", err)
+	}
+
+	uc, ok := claims[opts.ClaimUsername]
+	if ok {
+		res.Username = uc.(string)
+	}
+
+	gc := claims["groups"]
+	if gc != nil {
+		gif, ok := gc.([]interface{})
```

Comment:
I'd suggest delegating to the code in the auth package.

Technically, groups can be a _single string_ and at least one provider did this.

Reply:
Nice. This also slims down the code in this very package.